pkg/app/ops/insightcollector/collector.go

```diff
@@ -13,3 +13,353 @@
+		var a time.Time
+		var err error
+		switch kind {
+		case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
+			data, a, err = i.getInsightDataForDeployFrequency(ctx, appID, targetTimestamp, rangeFrom, to)
+		case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
+			data, a, err = i.getInsightDataForChangeFailureRate(ctx, appID, targetTimestamp, rangeFrom, to)
+		default:
+			return nil, 0, fmt.Errorf("invalid step: %v", kind)
+		}
+		if err != nil {
+			if err == ErrDeploymentNotFound {
+				if to.Equal(until) {
+					break
+				}
+				rangeFrom = to
+				to = movePoint(to, 1)
+				continue
+			}
+			return nil, 0, err
+		}
+
+		updatedps = append(updatedps, data)
+		rangeFrom = a
+		accumulatedTo = a
+	}
+
+	return updatedps, accumulatedTo.Unix(), nil
+}
+
+var (
+	ErrDeploymentNotFound = errors.New("deployments not found")
+)
+
+// getInsightDataForDeployFrequency accumulate insight data in target range for deploy frequency.
+func (i *InsightCollector) getInsightDataForDeployFrequency(
+	ctx context.Context,
+	applicationID string,
+	targetTimestamp int64,
+	from time.Time,
+	to time.Time) (*insightstore.DeployFrequency, time.Time, error) {
+	filters := []datastore.ListFilter{
+		{
+			Field:    "CreatedAt",
+			Operator: ">=",
+			Value:    from.Unix(),
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: "<",
+			Value:    to.Unix(),
+		},
+	}
+
+	if applicationID != "" {
+		filters = append(filters, datastore.ListFilter{
+			Field:    "ApplicationId",
+			Operator: "==",
+			Value:    applicationID,
+		})
+	}
+
+	deployments, err := i.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+		PageSize: pageSize,
+		Filters:  filters,
+	})
+	if err != nil {
+		i.logger.Error("failed to get deployments", zap.Error(err))
+		return &insightstore.DeployFrequency{}, time.Time{}, fmt.Errorf("failed to get deployments")
+	}
+	if len(deployments) == 0 {
+		return &insightstore.DeployFrequency{}, time.Time{}, ErrDeploymentNotFound
+	}
+
+	accumulatedTo := from.Unix()
+	for _, d := range deployments {
+		if d.CreatedAt > accumulatedTo {
+			accumulatedTo = d.CreatedAt
+		}
+	}
+
+	return &insightstore.DeployFrequency{
+		Timestamp:   targetTimestamp,
+		DeployCount: float32(len(deployments)),
+	}, time.Unix(accumulatedTo, 0).UTC(), nil
+}
+
+// getInsightDataForChangeFailureRate accumulate insight data in target range for change failure rate
+func (i *InsightCollector) getInsightDataForChangeFailureRate(
+	ctx context.Context,
+	applicationID string,
+	targetTimestamp int64,
+	from time.Time,
+	to time.Time) (*insightstore.ChangeFailureRate, time.Time, error) {
+
+	filters := []datastore.ListFilter{
+		{
+			Field:    "CreatedAt",
+			Operator: ">=",
+			Value:    from.Unix(),
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: "<",
+			Value:    to.Unix(),
+		},
+		{
+			Field:    "Status",
+			Operator: "in",
+			Value:    []model.DeploymentStatus{model.DeploymentStatus_DEPLOYMENT_FAILURE, model.DeploymentStatus_DEPLOYMENT_SUCCESS},
+		},
+	}
```

Comment:
Is this ok for the case:
- a deployment was created in the range [from, to] but it is still running while querying
- the deployment was completed successfully after this job

Will it be included in future jobs?

Reply:
Either pattern should be fine, since the status is not success or failure.

Is there any chance that the status will be changed from success or failure?

Reply:
OK,  use UpdatedAt instead of CreatedAt, thanks

Reply:
How about CANCELLED status?
Maybe we can just use `CompletedAt` instead of `UpdatedAt` + `Status`.

Reply:
> How about CANCELLED status?
Maybe we can just use CompletedAt instead of UpdatedAt + Status.

CompletedAt is zero value when CANCELLED status?

Reply:
No, `CompletedAt` will be set when the deployment is completed (SUCCESS | FAILURE | CANCELLED).

Reply:
OK

> How about CANCELLED status?

Did you mean `CANCELLED` should be included in the calculation of changeFailureRate?

Reply:
I am not sure. Just want to ask your thought. ww.
OK. Let's exclude them from the targets for this time.

Reply:
Ah, OK. 
Cancellation is a complicated concept to calculate because it feels like neither a success nor a failure, so I'm not sure too if it should be included in the calculation.

The way of calculation may change depending on whether we think of ChangeFailureRate as "how many failed deployments out of all the deployments that should have succeeded" or "how many failed deployments out of all the deployments that were attempted".

> OK. Let's exclude them from the targets for this time.

Agree. 👍 

So, I'll change only to use completedAt.